src/compiler/checker.ts

```diff
@@ -8228,6 +8358,27 @@ module ts {
                             break;
                     }
                 }
+
+                if (node.type) {
+                    if (languageVersion >= ScriptTarget.ES6 && isSyntacticallyValidGenerator(node)) {
+                        let returnType = getTypeFromTypeNode(node.type);
+                        if (returnType === voidType) {
+                            error(node.type, Diagnostics.A_generator_cannot_have_a_void_type_annotation);

```

Comment:
Well, you are allowed to write something that is an *iterator* or an *iterable* or both, or neither actually. You are allowed to write `{}`. You are just not allowed to write something like number because it has extra things that IterableIterator doesn't have.